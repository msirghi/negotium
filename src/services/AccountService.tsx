import Requests from '../common/requests/request';
import { BASE_API_URL_V1 } from '../common/constants/constants';
import { HttpMethod } from '../common/requests/types';
import ServiceResultFactory from '../common/requests/serviceResultFactory';

const getUserMetadata = () => {
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/metadata`,
    HttpMethod.GET
  )
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const updateUserTheme = (theme: string) => {
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/metadata`,
    HttpMethod.PATCH,
    { theme }
  )
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const AccountService = {
  getUserMetadata,
  updateUserTheme,
};

export default AccountService;
