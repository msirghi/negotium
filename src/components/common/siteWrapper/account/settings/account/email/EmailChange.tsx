import { Box } from '@mui/system';
import { Button, TextField } from '@mui/material';
import { FC, useState } from 'react';
import useTranslation from 'next-translate/useTranslation';
import ChevronLeftIcon from '@mui/icons-material/ChevronLeft';
import { Row } from '../../../../../utilities/row/Row';
import { RowDirection } from '../../../../../../../common/constants/enums';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../../../../../../redux/store';
import { useIsMobile } from '../../../../../../../common/hooks/common/useIsMobile';
import { useEmailChangeStyles } from './styles';
import { useFormik } from 'formik';
import ValidationService from '../../../../../../../services/ValidationService';
import { useSnackbar } from 'notistack';
import { LoadingButton } from '@mui/lab';
import AccountService from '../../../../../../../services/AccountService';
import { If } from '../../../../../utilities/if/If';
import { setUserEmail } from '../../../../../../../redux/account/accountSlice';

type Props = {
  onBackClick: () => void;
};

type FormValues = {
  newEmail: string;
  confirmNewEmail: string;
};

export const EmailChange: FC<Props> = ({ onBackClick }) => {
  const [error, setError] = useState('');
  const [loading, setLoading] = useState(false);

  const isMobile = useIsMobile();
  const classes = useEmailChangeStyles();
  const { enqueueSnackbar } = useSnackbar();
  const { t } = useTranslation('settings');

  const email = useSelector((state: RootState) => state.account.info.email);
  const dispatch = useDispatch();

  const onSubmit = async (values: FormValues) => {
    setError('');
    const errors = validateFields(values);
    if (Object.keys(errors).length !== 0) {
      return;
    }

    setLoading(true);
    try {
      const lowercasedEmail = values.newEmail.toLowerCase().trim();
      await AccountService.updateUserEmail(lowercasedEmail);
      enqueueSnackbar(t('change.emailUpdated'), { variant: 'success' });
      dispatch(setUserEmail(lowercasedEmail));
      formik.resetForm();
    } catch (e) {
      setError((e as Error).message);
    }
    setLoading(false);
  };

  const formik = useFormik({
    initialValues: {
      newEmail: '',
      confirmNewEmail: '',
    },
    onSubmit,
  });

  const validateFields = (values: FormValues) => {
    const errors: { [key: string]: string } = {};
    const { newEmail, confirmNewEmail } = values;
    const isValidEmail = ValidationService.isEmailValid(newEmail);
    if (!isValidEmail) {
      errors.newEmail = t('change.invalidEmail');
    }
    if (newEmail.toLowerCase() !== confirmNewEmail.toLowerCase()) {
      errors.confirmNewEmail = t('change.emailMismatch');
    }
    formik.setErrors(errors);
    return errors;
  };

  return (
    <Box className={classes.container}>
      <Box>
        <Button data-testid={'back-button'} startIcon={<ChevronLeftIcon />} onClick={onBackClick}>
          {t('change.back')}
        </Button>
      </Box>

      <Box className={classes.titleContainer}>
        <p className={classes.title}>{t('change.emailChangeTitle')}</p>
        <p className={classes.subtitle}>
          {t('change.currentEmail')}
          <b>{email}</b>
        </p>
      </Box>

      <form onSubmit={formik.handleSubmit}>
        <If condition={!!error}>
          <Box data-testid={'error-text'} className={classes.error}>
            {error}
          </Box>
        </If>
        <Row direction={RowDirection.COLUMN} className={classes.formContainer}>
          <TextField
            value={formik.values.newEmail}
            inputProps={{ 'data-testid': 'new-email-field' }}
            className={classes.input}
            onChange={formik.handleChange}
            name={'newEmail'}
            size={'small'}
            label={t('change.newEmail')}
            fullWidth={isMobile}
            error={formik.touched.newEmail && Boolean(formik.errors.newEmail)}
            helperText={formik.touched.newEmail && formik.errors.newEmail}
          />

          <TextField
            inputProps={{ 'data-testid': 'repeat-email-field' }}
            onChange={formik.handleChange}
            name={'confirmNewEmail'}
            value={formik.values.confirmNewEmail}
            fullWidth={isMobile}
            className={classes.input}
            size={'small'}
            label={t('change.confirmNewEmail')}
            error={
              formik.touched.confirmNewEmail &&
              Boolean(formik.errors.confirmNewEmail)
            }
            helperText={
              formik.touched.confirmNewEmail && formik.errors.confirmNewEmail
            }
          />
        </Row>

        <Row>
          <LoadingButton
            data-testid={'submit-button'}
            loading={loading}
            disabled={!formik.dirty}
            type={'submit'}
            fullWidth={isMobile}
            variant={'contained'}
            className={classes.saveButton}
            color={'primary'}
          >
            {t('change.save')}
          </LoadingButton>
        </Row>
      </form>
    </Box>
  );
};
