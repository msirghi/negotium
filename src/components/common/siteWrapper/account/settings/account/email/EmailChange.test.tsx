import { MockReduxProvider } from '../../../../../../../common/tests/TestUtils';
import { SnackbarProvider } from 'notistack';
import { EmailChange } from './EmailChange';
import renderer, { act } from 'react-test-renderer';
import { fireEvent, render, waitFor } from '@testing-library/react';
import AccountService from '../../../../../../../services/AccountService';

describe('EmailChange', () => {
  const defaultProps = {
    onBackClick: jest.fn(),
  };

  const renderComponent = () => {
    return (
      <MockReduxProvider
        reduxStore={{
          account: {
            info: { email: 'email@email.com' },
          },
        }}
      >
        <SnackbarProvider>
          <EmailChange {...defaultProps} />
        </SnackbarProvider>
      </MockReduxProvider>
    );
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const tree = renderer.create(renderComponent());
    expect(tree).toMatchSnapshot();
  });

  it('should handle back button click', () => {
    const { getByTestId } = render(renderComponent());
    const backButton = getByTestId('back-button');

    act(() => {
      fireEvent.click(backButton);
    });
    expect(defaultProps.onBackClick).toBeCalled();
  });

  it('should have disabled submit button by default', () => {
    const { getByTestId } = render(renderComponent());
    expect(getByTestId('submit-button')).toBeDisabled();
  });

  it('should submit the the form and show error if email is not valid', async () => {
    const { getByTestId, getByText } = render(renderComponent());
    const newEmailField = getByTestId('new-email-field');
    const submitButton = getByTestId('submit-button');

    act(() => {
      fireEvent.change(newEmailField, { target: { value: 'value' } });
      fireEvent.click(submitButton);
    });
    await waitFor(() => {
      expect(getByText('change.invalidEmail')).toBeInTheDocument();
    });
  });

  it('should submit the the form and show error if there is email mismatch', async () => {
    const { getByTestId, getByText } = render(renderComponent());
    const repeatEmailField = getByTestId('repeat-email-field');
    const submitButton = getByTestId('submit-button');

    act(() => {
      fireEvent.change(repeatEmailField, { target: { value: 'value' } });
      fireEvent.click(submitButton);
    });
    await waitFor(() => {
      expect(getByText('change.emailMismatch')).toBeInTheDocument();
    });
  });

  it('should make an api call to update the email is values are valid', async () => {
    const { getByTestId } = render(renderComponent());
    const repeatEmailField = getByTestId('repeat-email-field');
    const newEmailField = getByTestId('new-email-field');
    const submitButton = getByTestId('submit-button');
    const email = 'email@email.com';
    AccountService.updateUserEmail = jest.fn(() => Promise.resolve()) as any;

    act(() => {
      fireEvent.change(repeatEmailField, { target: { value: email } });
      fireEvent.change(newEmailField, { target: { value: email } });
      fireEvent.click(submitButton);
    });
    await waitFor(() => {
      expect(AccountService.updateUserEmail).toBeCalled();
    });
  });

  it('should make an api call to update the email is values are valid and case if different', async () => {
    const { getByTestId } = render(renderComponent());
    const repeatEmailField = getByTestId('repeat-email-field');
    const newEmailField = getByTestId('new-email-field');
    const submitButton = getByTestId('submit-button');
    AccountService.updateUserEmail = jest.fn(() => Promise.resolve()) as any;

    act(() => {
      fireEvent.change(repeatEmailField, {
        target: { value: 'email@email.com' },
      });
      fireEvent.change(newEmailField, { target: { value: 'email@email.COM' } });
      fireEvent.click(submitButton);
    });
    await waitFor(() => {
      expect(AccountService.updateUserEmail).toBeCalled();
    });
  });

  it('should make an api call to update the email is values are valid and show the error if something went wrong', async () => {
    const { getByTestId } = render(renderComponent());
    const repeatEmailField = getByTestId('repeat-email-field');
    const newEmailField = getByTestId('new-email-field');
    const submitButton = getByTestId('submit-button');
    const email = 'email@email.com';
    AccountService.updateUserEmail = jest.fn(() =>
      Promise.reject({ message: 'error' })
    ) as any;

    act(() => {
      fireEvent.change(repeatEmailField, { target: { value: email } });
      fireEvent.change(newEmailField, { target: { value: email } });
      fireEvent.click(submitButton);
    });
    await waitFor(() => {
      expect(getByTestId('error-text')).toBeInTheDocument();
    });
  });
});
