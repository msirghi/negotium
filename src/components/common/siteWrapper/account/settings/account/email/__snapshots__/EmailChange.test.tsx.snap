// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailChange should match the snapshot 1`] = `
<div
  className="makeStyles-container-1 MuiBox-root css-0"
>
  <div
    className="MuiBox-root css-0"
  >
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      data-testid="back-button"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronLeftIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
          />
        </svg>
      </span>
      change.back
    </button>
  </div>
  <div
    className="makeStyles-titleContainer-4 MuiBox-root css-0"
  >
    <p
      className="makeStyles-title-5"
    >
      change.emailChangeTitle
    </p>
    <p
      className="makeStyles-subtitle-6"
    >
      change.currentEmail
      <b>
        email@email.com
      </b>
    </p>
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="makeStyles-formContainer-2"
      onClick={[Function]}
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "flex-start",
          "width": "initial",
        }
      }
    >
      <div
        className="MuiFormControl-root MuiTextField-root makeStyles-input-3 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
        >
          change.newEmail
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
            data-testid="new-email-field"
            disabled={false}
            name="newEmail"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-dhh0px"
            >
              <span>
                change.newEmail
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root makeStyles-input-3 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
        >
          change.confirmNewEmail
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
            data-testid="repeat-email-field"
            disabled={false}
            name="confirmNewEmail"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-dhh0px"
            >
              <span>
                change.confirmNewEmail
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      onClick={[Function]}
      style={
        Object {
          "alignItems": "flex-start",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "width": "initial",
        }
      }
    >
      <button
        className="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled makeStyles-saveButton-7 css-1twrztz-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        data-testid="submit-button"
        disabled={true}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="submit"
      >
        change.save
      </button>
    </div>
  </form>
</div>
`;
